### DISCUSSION TITLE
Generate typescript definitions (d.ts)

### DISCUSSION THREAD

**Discussion by snebjorn** [fabricjs/fabric.js#7017] [Ideas]
What would it take to generate typescript definitions (d.ts) for fabric.js? (https://github.com/fabricjs/fabric.js/issues/6763#issuecomment-764972020)

According to https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html it looks simple. But after trying it out the d.ts file is empty. I believe it's because fabric.js is written in the IIFE style.

Does anyone know if IIFEs can be made to work?
Alternatively the code base could be converted to ES Module. But that's extra work and it might break stuff.

**Comment by asturur** [comment]
We can start from improving those:
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

Converting to es modules is a whole new journey that has less impact and for me has very low priority.

**Reply by snebjorn** [reply]
I played around with it and I don't think it's as simple as just improving the JSDoc - unfortunately :(

Problem is that typescript doesn't play nice with IIFEs.

I.E. it doesn't generate any types for the following
```js
(function (ns) {
    function foo() {
    }
    ns.foo = foo;
})(exports.ns || (exports.ns = {}));
```
It does however understand
```js
exports.ns = (() => {
    const ns = {};
    ns.foo = function () {

    }
    return ns;
})();
```
But that's CommonJS and it doesn't allow for exporting another `ns` that mutates it. Which fabric.js relies on.

So it doesn't look good.

**Comment by asturur** [comment]
could we take the typedef in the types library, fix it, and update it regularly?

**Comment by asturur** [comment]
i hope there was some kind of static analyzer of jsdocs, if jsdocs can create a full website from the doc notations it can also write a typedef file. I hoped someone created the template for it

**Comment by snebjorn** [comment]
Okay! I just needed to sleep on it :)
I think I found a way. But it will require a little refactoring.

See [TypeScript Playground Example](https://www.typescriptlang.org/play?useJavaScript=true#code/G4QwTgBAZiBGYEsDGEC8EDeBfA3AWACgZ5kA6YAUzAGcEB7AOzQgHIAWUjgBhfwMOKIkpAK4AXBABtSIBggC2IMRQDCdSXUjoAFFBEMkExtoCUmQhAgB6KxBUhJSEZKUUIsiAgYBaWBTEA7hQUTEjqmqQQAEr+ImAM1O4QAERgAOawIKbJENRiiAxppBbWtipgFAAmCGIAXBAAopUBXhAAsuASTAA8VS0MAAKwNXma8qSaaQB8JTaW8wsAFmJiAA61NgFbpMNio2Djk1YAVgCOIlQAnlZhGmDesgpK9Awn51cycorK3rea1KRjtQSnoDEZQg4nC5lGo7to-GkvAAaCAhSoo1Z0ahmDAlSygSB-LSsdKZbQsPELCAAaggq3A1AoAEkGGJtPCKIiGABtLgAXRpdKxEAAVBBtGjeQLvBAEV4pSYTCiAIxcMy0lhIikEKnzWn0mjM1nsuU85UC-XCsUShiVbnmiAy032vmKlVqwWa7W6yz6hlGtkcrncgBMFqFiWtkrDjtlnPlYbdEFVJj4lKJNPQXsFQdaADI86jbRAAPx0-0AMQ0SlzPIAzOHMZHxZKG7HnQ3FRB6srU+nwpBqVmTLxKRUxHFQgO+JYsIRZiKRXixSpFrI0hREmJFm4M1AwHR5BBGG4xHR3Aw6NuqBAWtvWkgqGIQK1VlR6JVj1AIBJ5BQUV4wB0AA1l4aQQEgDiSJkSDAYkICJKAkgXJ4iR+GBEFroUVTFDqlhigMf7yH4YAAPLfoIZDiFIy4QAMBogEeGAAMr5GBWDQAe8iwpoEAACo7rkz5gF04EZq0O4AB7HpApLaCAZhQGMSi4fMBEMUxrEFGkHFnjxkACW4eSdBh4lMFJMkQHJCnQMpYiqfhdEaZgAByIjEVQHHcpUcTPIwAoACK+eCX6YeubjaK08jYg5opOeAjGYKRsDHBQhheXQqzgtQAoAIJfH5TCZdltH0QlTEVvohgvBlWUvACjCruFAr2JI0EgLBOCeAwQHAVUx5MBQlBgJcEBIShSBYRupXORglVgjVEDcsV9UTAwajyKskj+BQLVQTBwFdYBIH9QEO5MONu5TW4CCJGEm3bcolQzeVmDzdVjC1dlpBEg0CFgQKf20IU0BVeCkQuVep5rmIP6LLdoMLYwA2SKNz59T+ATnuAaTuSEezik4YAVKyfEKP+EA+WAhUmJEAkiFugkVLjLiQI83wvKi-0g6CH0JBBsiXrDfgQAzOEvdTFVg4ty11YwAJwJoYh5ZU1Tgg4iN87eNSLBAGiIsIEBMt+46TluYAXCijWHltO2oQLbXi3hopWCC0vI+zrj6boXH6SiekDiiVOFVbcsJDilIEkJnT6cwDAUAE0BwEIpDe-uh76bTG5iMxdBxI+phIpSCxorH6Dx4nlHCN7Ad3Fn-i5-nFCF8X8yaAgXIOKR60249bjoCt8trRttvKEXztUu3neSN3TXYcwg8JMP10zm3YeJAP68QAAPtvmC4POE9V6IEjSJ7yi6MnVGnxMKVpfZFCScotraIv1AoriE+WMH4L1D-nO7wgAAVi4FwceupjIiQAGoOAuPUSBYg-atzRDA5CFB6il0Dq3WAlxUFwKLJUJBX8ubAzSPUXmIVtBExJmIMmf4UQILwRTHBTCg7BReBHYh+JwARiYQvdeP0BxAzAq3BYZY36CLuMIwoVC4g0LoRTf+jATCiPmD2WMHRtyCOoLI4m+MFHWEpuw5G1pNGLFIAABSZIYkMipV66lNvEB2UIvYDm0Iw2BzDcGeIxFiJhfZiFYHAVSOYa4tznnRjdJgHUwj6FhkpQkA4rKbjPNpexbce4PR2uQ92DBdE0JRJdCxVAkD+3JsUsASBOE+k8N+V+iBp6z17jtapNTkkTicVPLwXdMmj2bqohYkFHDOFcXCTBdwUTjM0CiNUwS2ljU8RU0pAz5i-goEslZASalzkCXMjJc8Nw5KRnk6h+NCmLJKWUv8SzWm6gQHUrpDAekHIoLcn09zxS5WJiAS4pBbpfOppcfJ+Muyf3mZYRxRUGndJnuta62gVmDMhCMmEbjTmshROisQMylSIvmEUy5eLLBrI2VwqkWy2k7PmY8558KsXnLQUsq56ySkUt1FSqkHLZx2JKDskokKLxPFRXcNMBAsAmFMDgIAA)

The IIFEs will have to assign their content like so
```js
fabric.util.animateColor = (function() {
  // omitted
  return animateColor;
})();
```

I was able to generate the following
```ts
declare namespace fabric {
    const version: string;
    namespace util {
        /**
         * Changes the color from one to another within certain period of time, invoking callbacks as value is being changed.
         * @memberOf fabric.util
         * @param {String} fromColor The starting color in hex or rgb(a) format.
         * @param {String} toColor The starting color in hex or rgb(a) format.
         * @param {Number} [duration] Duration of change (in ms).
         * @param {Object} [options] Animation options
         * @param {Function} [options.onChange] Callback; invoked on every value change
         * @param {Function} [options.onComplete] Callback; invoked when value change is completed
         * @param {Function} [options.colorEasing] Easing function. Note that this function only take two arguments (currentTime, duration). Thus the regular animation easing functions cannot be used.
         * @param {Function} [options.abort] Additional function with logic. If returns true, onComplete is called.
         */
        function animateColor(fromColor: string, toColor: string, duration?: number, options?: {
            onChange?: Function;
            onComplete?: Function;
            colorEasing?: Function;
            abort?: Function;
        }): void;
    }
}
```

**Comment by ShaMan123** [comment]
> +1
> @asturur I think `ts` must be included in this project for it to enhance DEV UX, make it easier to use and make it even more popular than it already is.
> The current definitions are lacking. I myself have a huge patch file for them and I don't bother to PR because it's too complicated IMO.
> I think that if it would be part of the project when creating a PR you can enforce the PR to update the `ts` files.
> I am sure the community will take part. I can say the least about myself!




I think this approach is a killer, going around instead of simply getting it done.
@asturur I understand you don't use `ts` (I have no idea why 😯). It is very similar to JSDOC and has many benefits (like the build process).
I do think that if you want this amazing project to be more accessible and easy to use it is a must.
Another thought, an open source project must evolve for and from the community.
So to summarize I think that adding `.d.ts` files is the solution.
And again, I believe the community will help, I know I will.

**Reply by asturur** [reply]
Adding a d.ts file is what we want to do indeed.

**Reply by asturur** [reply]
Also i do not use d.ts because my day to day work is in JS, and this library is in JS.
So i just do not have occasion to use it regularly, and when i do not work or code fabricJS i do things unrelated to programming.

**Comment by asturur** [comment]
If you want to move on on this, we need someone that pick up an existiing t.ds file, delete the wrong parts that are not compatible with 4.5.0 and commits it.
Then we neeed a list of things to fix and we can start fixing it.

**Reply by ShaMan123** [reply]
I've started a [draft](https://github.com/ShaMan123/fabric.js/tree/types).
My approach is to create a `types` folder as the exact copy of `src`. I think it will make maintaining it much easier, as opposed to a huge `index.d.ts` file. 
The second advantage is that this approach doesn't use `namespace`, which is not recommended by the spec. This is import because of the `extend` approach of `fabric`, types must be esaily `extended` and `namespace` makes it harder I believe. 
I have set it up in terms of configuration so now it's ready for work.

**Reply by ShaMan123** [reply]
@snebjorn How did you manage to emit the declaration files? I didn't understand.

**Reply by ShaMan123** [reply]
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

**Reply by snebjorn** [reply]
> @snebjorn How did you manage to emit the declaration files? I didn't understand.

Here's my config, with that I just ran `tsc`
```jsonc
// tsconfig.json
{
  // Change this to match your project
  "include": ["dist/fabric.js"],
  "compilerOptions": {
    // Tells TypeScript to read JS files, as
    // normally they are ignored as source files
    "allowJs": true,
    // Generate d.ts files
    "declaration": true,
    // This compiler run should only output d.ts files
    "emitDeclarationOnly": true,
    // Types should go into this directory.
    // Removing this would place the .d.ts files
    // next to the .js files
    "outDir": "types"
  }
}
```

But it'll require changes to the js files mention here https://github.com/fabricjs/fabric.js/discussions/7017#discussioncomment-626265

**Reply by bobbhy** [reply]
Any updates please 🙏 I'm using right now version 5.3.1 in my end of year project, and I'm not into creating another JS project alongside a huge angular app for a simple canva.com like editor, I have tried using tools like dts-gen and npm-dts for d.ts generation and I have also followed this guide https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html using a clone of fabric JS repo without any success I get endless errors, and I'm really stuck right now 😔.

**Comment by ShaMan123** [comment]
#7596
